ScarletEditor/src/Panels/OutputLog.cpp

```diff
@@ -0,0 +1,122 @@
+#include "Panels/OutputLog.h"
+
+namespace ScarletEngine
+{
+	OutputLogPanel::OutputLogPanel()
+		: FilterText()
+		, CurrentMessageIndex(0)
+		, NumMessages(0)
+	{
+		
+		MessageBuffer.resize(MaxBufferLength);
+		FilterText.reserve(32);
+
+		Logger::Get().GetOnMessageLogged().Bind(this, &OutputLogPanel::OnMessageLogged);
+	}
+
+	OutputLogPanel::~OutputLogPanel()
+	{
+		Logger::Get().GetOnMessageLogged().Unbind(this);
+	}
+
+	void OutputLogPanel::Draw()
+	{
+		if (ImGui::Begin("Output Log"))
+		{
+			ImGui::TextUnformatted(ICON_MD_FILTER_LIST " Filter:");
+			ImGui::SameLine();
+			ImGui::InputText("###Filter", (char*)FilterText.c_str(), FilterText.capacity(), ImGuiInputTextFlags_EnterReturnsTrue);
+			
+			ImGui::SameLine();
+
+			if (ImGui::Button("Clear"))
+			{
+				Clear();
+			}
+			ImGui::SameLine();
+
+			if (ImGui::Button(ICON_MD_SETTINGS))
+			{
+				ImGui::OpenPopup("###OutputSetting");
+			}
+
+			if (ImGui::BeginPopup("###OutputSetting"))
+			{
+				ImGui::MenuItem("Show Verbose", "", &bShowVerbose);
+				ImGui::MenuItem("Show Info", "", &bShowInfo);
+				ImGui::MenuItem("Show Warning", "", &bShowWarning);
+				ImGui::MenuItem("Show Error", "", &bShowError);
+				ImGui::Separator();
+				ImGui::MenuItem("Autoscroll", "", &bAutoscroll);
+				ImGui::EndPopup();
+			}
+
+			if (ImGui::BeginChild("Log"))
```

Comment:
Not too familiar with immediate mode gui, so you have to redraw every single message every frame?

Reply:
Yeah I'm not sure there doesn't _seem_ to be a better way to do it. The ImGUI demo log does a similar thing and redraws every frame. There are some fancy things you can do with `ImGuiListClipper` which I can look into uses for a future revision of the logger. There are also some performance improvements which could be had by caching the results of `PassesFilter` rather than checking each message every frame.

Current just drawing the text takes ~0.5 microseconds per message. If we include `PassesFilter`, it goes up to ~0.8 microseconds (with some profiling overhead).